data-pipeline/src/trace_exporter.rs

```diff
@@ -214,32 +253,173 @@ impl TraceExporter {
+                let stats_status: Option<StatsComputationStatus> =
+                    std::sync::Arc::<StatsComputationStatus>::into_inner(
+                        self.client_side_stats.into_inner(),
+                    );
+                if let Some(StatsComputationStatus::Enabled {
+                    stats_concentrator: _,
+                    cancellation_token,
+                    exporter_handle,
+                }) = stats_status
+                {
+                    cancellation_token.cancel();
+                    let _ = exporter_handle.await;
+                }
+            });
+            Ok(())
+        }
+    }
+
+    /// Start the stats exporter and enable stats computation
+    ///
+    /// Should only be used if the agent enabled stats computation
+    fn start_stats_computation(
+        &self,
+        span_kinds: Vec<String>,
+        peer_tags: Vec<String>,
+    ) -> anyhow::Result<()> {
+        if let StatsComputationStatus::DisabledByAgent { bucket_size } =
+            **self.client_side_stats.load()
+        {
+            let stats_concentrator = Arc::new(Mutex::new(SpanConcentrator::new(
+                bucket_size,
+                time::SystemTime::now(),
+                span_kinds,
+                peer_tags,
+            )));
+
+            let cancellation_token = CancellationToken::new();
+
+            let mut stats_exporter = stats_exporter::StatsExporter::new(
+                bucket_size,
+                stats_concentrator.clone(),
+                self.metadata.borrow().into(),
+                Endpoint::from_url(add_path(&self.endpoint.url, STATS_ENDPOINT)),
+                cancellation_token.clone(),
+            );
+
+            let exporter_handle = self.runtime.spawn(async move {
+                stats_exporter.run().await;
+            });
+
+            self.client_side_stats
+                .store(Arc::new(StatsComputationStatus::Enabled {
+                    stats_concentrator,
+                    cancellation_token,
+                    exporter_handle,
+                }));
+        };
+        Ok(())
+    }
+
+    /// Stops the stats exporter and disable stats computation
+    ///
+    /// Used when client-side stats is disabled by the agent
+    fn stop_stats_computation(&self) {
+        if let StatsComputationStatus::Enabled {
+            stats_concentrator,
+            cancellation_token,
+            exporter_handle: _,
+        } = &**self.client_side_stats.load()
+        {
+            self.runtime.block_on(async {
+                cancellation_token.cancel();
+            });
+            let bucket_size = stats_concentrator.lock().unwrap().get_bucket_size();
+            self.client_side_stats
+                .store(Arc::new(StatsComputationStatus::DisabledByAgent {
+                    bucket_size,
+                }));
+        }
+    }
+
+    /// Check for a new state of agent_info and update the trace exporter if needed
+    fn check_agent_info(&self) {
```

Comment:
Is it possible that this (and other large functions as part of this PR) be broken up into smaller ones to make them easier to read and maintain?